packages/workspace/src/utilities/generate-globs.ts

```diff
@@ -39,6 +40,26 @@ export function createGlobPatternsForDependencies(
       projectGraph
     );
 
+    const dirsToUse = [];
+    const recursiveScanDirs = (dirPath) => {
+      const dirs = readdirSync(dirPath)
+        .map((d) => join(dirPath, d))
+        .filter((d) => lstatSync(d).isDirectory());
+      for (const dir of dirs) {
+        if (readdirSync(dir).includes('ng-package.json')) {
+          dirsToUse.push(dir);
+        } else {
+          recursiveScanDirs(dir);
+        }
+      }
```

Comment:
We need to make sure we only iterate over non-ignored files, so we should do something similar to what's in the `visitNotIgnoredFiles` helper (that's for generators).

Also, let's reduce the number of loops and array allocations (`map` and `filter` will loop and allocate new arrays). We could do everything in a single loop and avoid reading all children inside a directory to know if a file exists:

```suggestion
      const children = readdirSync(dirPath);
      for (const child of chlidren) {
        if (!lstatSync(child).isDirectory()) {
          continue;
        }
        const dir = join(dirPath, child);
        if (existsSync(join(dir, 'ng-package.json'))) {
          dirsToUse.push(dir);
        } else {
          recursiveScanDirs(dir);
        }
      }
```

Note the above doesn't include the non-ignored files filter I mentioned. It can be added to the already existing check for `!lstatSync(child).isDirectory()`.

Reply:
We’ll still need the map here to build the path. 
```
const children = readdirSync(dirPath);
      for (const child of chlidren) {
        if (!lstatSync(child).isDirectory()) {
```

because `readdirSync` only returns the name of file or directory, not the full path, so lstatSync(child) will fallover.

but good point on the other optimisations and the ignored files

Reply:
You don't need the `map`, you just need to `join` inside the loop. What I put in line 61, moves to line 58, before checking if it's a directory.

Reply:
The point is that while `map`, `filter` and the likes are easier to read, they come with a cost. That cost might not be too important in certain scenarios where the collections are finite and small, but in others, like traversing the filesystem (especially when running as part of a task where perf matters), it's important and we can always achieve the same without using them.